server-spi-private/src/main/java/org/keycloak/models/utils/SessionExpirationUtils.java

```diff
@@ -0,0 +1,203 @@
+     * @param realm The realm model
+     * @return The time when the user session is expired or -1 if does not expire
+     */
+    public static long calculateUserSessionMaxLifespanTimestamp(boolean offline, boolean isRememberMe, long created, RealmModel realm) {
+        long timestamp = -1;
+        if (offline) {
+            if (realm.isOfflineSessionMaxLifespanEnabled()) {
+                timestamp = created + TimeUnit.SECONDS.toMillis(getOfflineSessionMaxLifespan(realm));
+            }
+        } else {
+            long userSessionMaxLifespan =  TimeUnit.SECONDS.toMillis(getSsoSessionMaxLifespan(realm));
+            if (isRememberMe) {
+                userSessionMaxLifespan = Math.max(userSessionMaxLifespan, TimeUnit.SECONDS.toMillis(realm.getSsoSessionMaxLifespanRememberMe()));
+            }
+            timestamp = created + userSessionMaxLifespan;
+        }
+        return timestamp;
+    }
+
+    /**
+     * Calculates the time in which the user session is expired via the idle
+     * configuration.
+     * @param offline is the session offline?
+     * @param isRememberMe is the session remember me?
+     * @param lastRefreshed The last time the session was refreshed
+     * @param realm The realm model
+     * @return The time in which the user session is expired by idle timeout
+     */
+    public static long calculateUserSessionIdleTimestamp(boolean offline, boolean isRememberMe, long lastRefreshed, RealmModel realm) {
+        long timestamp;
+        if (offline) {
+            timestamp = lastRefreshed + TimeUnit.SECONDS.toMillis(getOfflineSessionIdleTimeout(realm));
+        } else {
+            long userSessionIdleTimeout = TimeUnit.SECONDS.toMillis(getSsoSessionIdleTimeout(realm));
+            if (isRememberMe) {
+                userSessionIdleTimeout = Math.max(userSessionIdleTimeout, TimeUnit.SECONDS.toMillis(realm.getSsoSessionIdleTimeoutRememberMe()));
+            }
+            timestamp = lastRefreshed + userSessionIdleTimeout;
+        }
+        return timestamp;
+    }
+
+    /**
+     * Calculates the time in which the client session is expired via lifespan
+     * configuration in the realm and client.
+     * @param offline is the session offline?
+     * @param isRememberMe is the session remember me?
+     * @param clientSessionCreated timestamp when the client session was created
+     * @param userSessionCreated timestamp when the user session was created
+     * @param realm The realm model
+     * @param client The client model
+     * @return The time when the client session is expired or -1 if does not expire
+     */
+    public static long calculateClientSessionMaxLifespanTimestamp(boolean offline, boolean isRememberMe,
+            long clientSessionCreated, long userSessionCreated, RealmModel realm, ClientModel client) {
+        long timestamp = -1;
+        if (offline) {
+            if (realm.isOfflineSessionMaxLifespanEnabled()) {
```

Comment:
Let's assume that:
- Realm settings have "Offline Session Max Lifespand" disabled
- Client has `Client Offline Token Max` overriden to 1 day.

IMO the correct offline client session timeout should be 1 day with those settings? However it looks that current logic in this PR will just use `never` (or `-1`) with those settings due this "if" check, which causes that disabled `Offline Session Max Lifespan` won't take into account overriden values for `Client Offline Token Max` ?

Reply:
I prefer to maintain this for the moment. It was like this in both implementations. In order to have expiration for offline session the realm `isOfflineSessionMaxLifespanEnabled` should be enabled first. If not offline sessions never expire by lifespan.

Reply:
Yes, that works for me.

However it will be good to improve the usability around this IMO, but that can be done as a follow-up. Currently if I configure the things in the admin console as I've mentioned above (`Offline Session Max Lifespan` disabled for the realm, but `Client Offline Token Max` overriden to 1 day for client `foo`), I would expect that offline client session for my client `foo` would expire in 1 day.

The possibility here is maybe to completely hide the `Client Offline Token Max` switch for the clients in case that realm `Offline Session Max Lifespan` is disabled. As the `Client Offline Token Max` client switch in-fact doesn't have any effect in this case and hence exposing it in the admin console is confusing IMO. But I see that as a possible follow-up...